src/py/flwr/common/retry_invoker.py

```diff
@@ -0,0 +1,287 @@
+) -> Generator[float, None, None]:
+    """Wait time generator for specified intervals.
+
+    Parameters
+    ----------
+    interval: Union[float, Iterable[float]] (default: 1)
+        A constant value to yield or an iterable of such values.
+    """
+    if not isinstance(interval, Iterable):
+        interval = itertools.repeat(interval)
+    yield from interval
+
+
+def full_jitter(max_value: float) -> float:
+    """Randomize a float between 0 and the given maximum value.
+
+    This function implements the "Full Jitter" algorithm as described in the
+    AWS article discussing the efficacy of different jitter algorithms.
+    Reference: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+
+    Parameters
+    ----------
+    max_value : float
+        The upper limit for the randomized value.
+    """
+    return random.uniform(0, max_value)
+
+
+@dataclass
+class RetryState:
+    """State for callbacks in RetryInvoker."""
+
+    target: Callable[..., Any]
+    args: Tuple[Any, ...]
+    kwargs: Dict[str, Any]
+    tries: int
+    elapsed_time: float
+    exception: Optional[Exception] = None
+    actual_wait: Optional[float] = None
+
+
+# pylint: disable-next=too-many-instance-attributes
+class RetryInvoker:
+    """Wrapper class for retry (with backoff) triggered by exceptions.
+
+    Parameters
+    ----------
+    wait_strategy: Generator[float, None, None]
+        A generator yielding successive wait times in seconds. If the generator
+        is finite, the giveup event will be triggered when the generator raises
+        `StopIteration`.
+    exception: Union[Type[Exception], Tuple[Type[Exception]]]
+        An exception type (or tuple of types) that triggers backoff.
+    max_tries: Optional[int]
+        The maximum number of attempts to make before giving up. Once exhausted,
+        the exception will be allowed to escape. If set to None, there is no limit
+        to the number of tries.
+    max_time: Optional[float]
+        The maximum total amount of time to try before giving up. Once this time
+        has expired, this method won't be interrupted immediately, but the exception
+        will be allowed to escape. If set to None, there is no limit to the total time.
+    jitter: Optional[Callable[[float], float]] (default: full_jitter)
+        A function of the value yielded by `wait_strategy` returning the actual time
+        to wait. This function helps distribute wait times stochastically to avoid
+        timing collisions across concurrent clients. Wait times are jittered by
+        default using the `full_jitter` function. To disable jittering, pass
+        `jitter=None`.
+    should_giveup: Optional[Callable[[Exception], bool]] (default: None)
+        A function accepting an exception instance, returning whether or not
+        to give up prematurely before other give-up conditions are evaluated.
+        If set to None, the strategy is to never give up prematurely.
```

Comment:
These should go last to align with the order in `__init__`

Reply:
fixed. thanks for pointing it out.